File: 프로그래머스/lv2/42626. 더 맵게/더 맵게.java
/*

    k<==1,000,000,000 -> int 범위
    스코빌 <= 1,000,000 ->nlogn 알고리즘 필요
    
    작은 것부터 연산 -> 정렬 -> 우선순위 큐
*/

import java.util.*;

class Solution {
    public int solution(int[] scoville, int K) {
        int answer = 0;
        
        PriorityQueue<Integer> scovillePQ = new PriorityQueue<>();
        
        for(int val : scoville){
            scovillePQ.add(val);
        }
        
        while(scovillePQ.size()>1){
            int first = scovillePQ.poll();
            int second = scovillePQ.poll();
            
            if(first >=K){
                break;
            }
            
            int sum = first + second*2;
            scovillePQ.add(sum);
            
            answer++;
        }
        if(scovillePQ.size()==1 && scovillePQ.peek() < K){
            return -1;
        }
        
        return answer;
    }
}

File: 프로그래머스/2/42626. 더 맵게/더 맵게.py
from queue import PriorityQueue

def solution(scoville, K):
    answer = 0
    pq = PriorityQueue()
    
    for s in scoville:
        pq.put(s)
    cnt = 0
    
    while pq.qsize() > 1:
        first = pq.get()
        
        if first >= K:
            return cnt
        
        second = pq.get()
        
        pq.put(first + second * 2)
        cnt = cnt + 1
        
    if pq.get() <= K:
        return -1
        
    return cnt